opencart 2.2.0.0/upload/admin/model/smailyforopencart/admin.php

```diff
@@ -18,13 +18,66 @@ public function install() {
         );
     }
 
+
+    public function upgrade() {
+        // Load models.
+        $this->load->model('extension/modification');
+        $this->load->model('setting/setting');
+        // Module modification settings.
+        $smaily_module = $this->model_extension_modification->getModificationByCode('smaily_for_opencart_extension');
+        
+        // Stop if module hasn't been installed with adding it as a modification.
+        if (empty($smaily_module)) {
+            return;
+        }
+
+        $version = $smaily_module['version'];
+
+        // Version 1.2.0 - standardize abandoned cart fields.
+        if (version_compare($version, '1.2.0', '=')) {
+            $cart_additional = $this->config->get('smaily_for_opencart_abandoned_additional');
+
```

Comment:
I have been working on this upgrade function for a while. But I have reached a situation where I think it is impossible to create such a script that works.

If our plugin is installed with adding it as a modification then there is a version we can compare with.
The issue is that the same code can't be used again to upload the same plugin with upgraded version. That makes version compare redundant as we would have to know the specific code that comes with that version.
![image](https://user-images.githubusercontent.com/22918359/74733608-5893cb00-5255-11ea-90bb-8f95d698316a.png)

So maybe one option would be to include the plugin version in the code but then it's the case where two identical plugins are shown. https://github.com/sendsmaily/smaily-opencart-module/commit/11e4c27a0f575f99be78f959902e770f70e1dd2c

Currently, it seems to me as the opencart logic is that you have to uninstall your plugin to install the new version .

@TomAbel @kaittodesk maybe you have some good ideas.

Reply:
So if we keep the code the same that means customers cant upload a new plugin before uninstalling the last one.

If we change the code and add a version to it they can add a new plugin but have both of them installed.

From both of these bad options, I would pick the first one.

Reply:
Keep code same, it's OC fault they can't upload a new plugin before uninstalling ¯\_(ツ)_/¯

Reply:
I concur. If OpenCart does not have a proper upgrade path for modules then uninstall-install is the way to go.

Reply:
Okay, after messing around a while I think I figured it out.

The module modification can be removed without uninstalling the module. Kind of weird but it just removes your module from the `oc_modifications` table. That makes possible to install plugin with the same code again.

I created a hook that runs after each plugin update and updates our plugin if it's necessary. https://github.com/sendsmaily/smaily-opencart-module/pull/29/commits/9bb9c8206482bcd96849628b8da9d142b94269aa

For me this workflow it's weird - you have to remove modification and then install it again. But maybe it's how opencart should work and admins use this functionality.

This script won't run for our old plugins because they don't have this update hook registered (added with installation) but I would keep it as it is because this provides us better options for the future.

I will add also a notice that customers should uninstall older version and install new or add our upgrade hook to their database.